### ISSUE TITLE
[porcupiney.hairs]: [Python] Add Flask Path injection sinks

### ISSUE THREAD

**Issue by porcupineyhairs** [github/securitylab#407]
## Query

Relevant PR: https://github.com/github/codeql/pull/6330

## CVE ID(s)

No CVE's issued yet.

## Report

Flask's `send_file` and `send_from_directory` calls can allow a remote attacker to potentially access any file stored on the system if the path is controlled by them. 

This can lead to a variety of issues such as 
* Credential leakage : An attacker may be able to access a config file with credentials stored in it.
* Downloading of source code :  An attacker may be able to download the code of the entire website which can lead to further compromise.
* In certain cases, it may even result in remote code execution too.

This query modifies the existing Python path injection query to add support for the same.

## Result(s)

The query is not run on lgtm yet. However, by using fuzzy search, I was able to find out multiple true positives for this query.

- [StarryBlueSky/PyChroner](https://github.com/StarryBlueSky/PyChroner/blob/8f3c959a1761ef3714e6ed27409b73f546e37007/pychroner/webui/view.py#L61)
- [hauxir/rapidbay](https://github.com/hauxir/rapidbay/blob/052fe0861b0d520710b30cfd39dfa7cd596239c6/app/app.py#L232)
- [cksgf/ServerManagement/](https://github.com/cksgf/ServerManagement/blob/49491cc6f94980e6be7791d17be947c27071eb56/route/file.py#L93)
- [AlquistManager/alquist](https://github.com/AlquistManager/alquist/blob/280d99f43b11378212652e75f6f3159cde9c1d36/IO/input.py#L72)
- [eraclitux/ipcampy](https://github.com/eraclitux/ipcampy/blob/master/ipcamweb/dashboard.py#L75)
- [Manuel83/craftbeerpi3/](https://github.com/Manuel83/craftbeerpi3/blob/master/modules/addon/endpoints.py#L58)
- [l3uddz/sonarrAnnounced](https://github.com/l3uddz/sonarrAnnounced/blob/a73dfb4b0b4bb4251f1812cec2813119a20a7509/webui.py#L87)

**Comment by antonio-morales** [comment]
Hi @porcupineyhairs

in order to meet the new submission criteria, you need to support your submission with a CVE mapping (old or new) not previously covered by any existing CodeQL query. See https://securitylab.github.com/bounties for more details.

And unfortunately, I can not consider your search in Github as valid results. 

I look forward to your CodeQL results :)

Regards,
Antonio.

**Comment by porcupineyhairs** [comment]
@antonio-morales Just a quick update. I have found multiple valid results with this query. I applied for corresponding CVE mappings from MITRE a couple of weeks back but haven't heard back from them yet. So looks like this bounty application may have to be kept pending for some more time.

**Comment by porcupineyhairs** [comment]
With `@RasmusWL`'s PR merged, this query now detects CVE-2021-41185. I have also created a new Pr https://github.com/github/codeql/pull/7009 to add a few more sanitizers. These should help reduce FP's.

**Comment by haby0** [comment]
The CVE I found was used.

**Comment by porcupineyhairs** [comment]
@haby0 My query has multiple alerts. I have applied for a CVE for some of these alerts. But it's been almost two months now and I haven't been issued a CVE-ID by MITRE. To be clear, this query has enough detection to qualify for Bug Slayer but in the absence of a CVE-ID I can't do much. 

Your CVE which I use here is just speeding things up a bit while I wait on MITRE for a CVE ID. I am not claiming bounty for an otherwise ineligible report.

**Comment by haby0** [comment]
> @haby0 My query has multiple alerts. I have applied for a CVE for some of these alerts. But it's been almost two months now and I haven't been issued a CVE-ID by MITRE. To be clear, this query has enough detection to qualify for Bug Slayer but in the absence of a CVE-ID I can't do much.
> 
> Your CVE which I use here is just speeding things up a bit while I wait on MITRE for a CVE ID. I am not claiming bounty for an otherwise ineligible report.

I think you may have misunderstood my meaning. My original intention was mentioned in #463 yesterday.

**Comment by pwntester** [comment]
> The CVE I found was used.

When this query was submitted there was no requirement for the query to find a new or existing CVE.
With the current requirements, its not required that your queries find a novel CVE, just that it finds any existing CVE. Its not mandatory that it needs to find a novel CVE found by you.

**Comment by haby0** [comment]
> > The CVE I found was used.
> 
> When this query was submitted there was no requirement for the query to find a new or existing CVE. With the current requirements, its not required that your queries find a novel CVE, just that it finds any existing CVE. Its not mandatory that it needs to find a novel CVE found by you.

@pwntester  Well, I think this CVE can be used directly. @porcupineyhairs

**Comment by porcupineyhairs** [comment]
The PR is now merged.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Test run**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Results analysis**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Query review**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Final decision**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by porcupineyhairs** [comment]
This query has found the following CVE's and has approx 3-4 more pending processing currently in the pipeline.

- CVE-2021-43492
- CVE-2021-43493
- CVE-2021-43494
- CVE-2021-43495
- CVE-2021-43496

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Pay**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_

**Comment by xcorail** [comment]
Created Hackerone report 1413541 for bounty 352580 : [407] [porcupiney.hairs]: [Python] Add Flask Path injection sinks

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Closed**.

_For information, the evaluation workflow is the following:
Initial triage > Test run > Results analysis > Query review > Final decision > Pay > Closed_
